### PR TITLE
Support custom class mappings. Closes #233

### LINKED ISSUE #233: Support CSS Modules

### PR DESCRIPTION
**Pull request by tadatuta**
cc @miripiruni @veged @zxqfox

Closes #233

### CODE REVIEW

File: lib/bemhtml/index.js [hunk 1]
```diff
@@ -15,6 +16,14 @@ module.exports = BEMHTML;
 
 BEMHTML.prototype.Entity = Entity;
 
+BEMHTML.prototype._mapClass = function _mapClass(classname, separator) {
+  if (!this._classMap) return classname;
+
+  return separator ?
+    classname.split(' ').map(this._mapClass, this).join(' ') :
```
Comment (zxqfox) [inline lib/bemhtml/index.js]: `split(separator)`

File: lib/bemhtml/index.js [hunk 2]
```diff
@@ -312,8 +321,8 @@ BEMHTML.prototype.buildModsClasses = function buildModsClasses(block,
 
     var builder = this.classBuilder;
     res += ' ' + (elem ?
-                  builder.buildElemClass(block, elem, modName, modVal) :
-                  builder.buildBlockClass(block, modName, modVal));
+        this._mapClass(builder.buildElemClass(block, elem, modName, modVal)) :
+        this._mapClass(builder.buildBlockClass(block, modName, modVal)));
```
Comment (zxqfox) [inline lib/bemhtml/index.js]: может вокруг `(elem ? : )` ?

File: test/options-classmap-test.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      block: 'b1',
+      elem: 'e1',
+      mix: [
+          { block: 'b2' },
+          { block: 'b3', mods: { m1: 'v1' } }
+      ]
+    }), '<div class="ololo bedva betri betrim1v1"></div>');
+
+    template = bemxjst.compile('', { classMap: {
+      b1: 'ololo',
```
Comment (zxqfox) [inline test/options-classmap-test.js]: Guess, it's redundant.

**Comment by zxqfox** [comment]
LGTM

**Comment by miripiruni** [comment]
Nice feature, @tadatuta!

We need: 
 * [ ] Documentation.
 * [ ] Several small test cases instead of big one.
 * [ ] Benchmark results. (Without option mainly).

**Comment by tadatuta** [comment]
can we trust benchmark results ATM?

**Comment by miripiruni** [comment]
@tadatuta yes. More info: https://github.com/bem/bem-xjst/blob/escaping2/bench/README.md

File: lib/bemhtml/index.js [hunk 4]
```diff
@@ -8,13 +8,22 @@ function BEMHTML(options) {
...
   this._shortTagCloser = xhtml ? '/>' : '>';
+  this._classMap = options.classMap;
 }
 
 inherits(BEMHTML, BEMXJST);
 module.exports = BEMHTML;
 
 BEMHTML.prototype.Entity = Entity;
 
+BEMHTML.prototype._mapClass = function _mapClass(classname, separator) {
```
Comment (tadatuta) [inline lib/bemhtml/index.js]: `\s+`

Comment (miripiruni) [inline lib/bemhtml/index.js]: `className`

File: lib/bemhtml/index.js [hunk 5]
```diff
@@ -8,13 +8,22 @@ function BEMHTML(options) {
 
   var xhtml = typeof options.xhtml === 'undefined' ? true : options.xhtml;
   this._shortTagCloser = xhtml ? '/>' : '>';
+  this._classMap = options.classMap;
```
Comment (tadatuta) [inline lib/bemhtml/index.js]: classMap -> classesMap

File: lib/bemhtml/index.js [hunk 6]
```diff
@@ -311,9 +320,9 @@ BEMHTML.prototype.buildModsClasses = function buildModsClasses(block,
       modVal += '';
 
     var builder = this.classBuilder;
-    res += ' ' + (elem ?
-                  builder.buildElemClass(block, elem, modName, modVal) :
-                  builder.buildBlockClass(block, modName, modVal));
+    res += ' ' + this._mapClass((elem ?
```
Comment (tadatuta) [inline lib/bemhtml/index.js]: extra brackets

File: lib/bemhtml/index.js [hunk 7]
```diff
@@ -8,13 +8,22 @@ function BEMHTML(options) {
...
+  this._classMap = options.classMap;
 }
 
 inherits(BEMHTML, BEMXJST);
 module.exports = BEMHTML;
 
 BEMHTML.prototype.Entity = Entity;
 
+BEMHTML.prototype._mapClass = function _mapClass(classname, separator) {
+  if (!this._classMap) return classname;
```
Comment (tadatuta) [inline lib/bemhtml/index.js]: classname -> className

**Comment by miripiruni** [comment]
What current status here?

**Comment by tadatuta** [comment]
Nothing changed since we talked last time. I'll let you know when it'd be updated.

**Comment by vithar** [comment]
I need this feature in granny serp. @miripiruni can you continue this task by yourself?

**Comment by veged** [comment]
@vithar I have a strong opinion against such classNames minification — it brings more problems than profit (as we experienced in past at SERP)

and it's definitely non-blocker feature for Granny SERP

**Comment by vithar** [comment]
Granny have no js, in this case there is no problem with class minification. I'll do it anyway.

And yes, this is not show stopper.

File: .jshintrc [hunk 8]
```diff
@@ -6,7 +6,7 @@
 
     // Enforcing
     "bitwise"       : false,     // true: Prohibit bitwise operators (&, |, ^, etc.)
-    "camelcase"     : true,      // true: Identifiers must be in camelCase
+    "camelcase"     : false,      // true: Identifiers must be in camelCase
```
Comment (zxqfox) [inline .jshintrc]: Better to check this in JSCS/Eslint

**Comment by miripiruni** [comment]
@vithar Am I right that class minification is useless? https://nda.ya.ru/3S2rwk And may I close this PR?

**Comment by tadatuta** [comment]
it's not just about minification it's also about possibility to use CSS modules with `bem-xjst`

**Comment by miripiruni** [comment]
@tadatuta OK.

**Comment by vithar** [comment]
Also class minification can be used to hide code from adblock

**Pull request closed without merging**